src/v/cloud_storage/remote_partition.cc

```diff
@@ -647,18 +646,25 @@ std::unique_ptr<remote_segment_batch_reader> remote_partition::borrow_reader(
   storage::log_reader_config config, model::offset key, segment_state& st) {
     if (std::holds_alternative<offloaded_segment_state>(st)) {
         gc_stale_materialized_segments(true);
+    } else if (_readers_limit.current() <= 0) {
```

Comment:
Looking at ss semaphore implementation, `current()` is always `>= 0`. If we want to check `<= 0` we should use `available_units()`